File: downloads/code/LeetCode/Python/Perfect-Squares.py
import sys
class Solution(object):
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        # dp[i+j*j] = min(dp[i]+1, dp[i+j*j])

        dp = [sys.maxint]*(n+1)
        i = 1
        while i*i <= n:
            dp[i*i] = 1
            i += 1
        for i in range(1,n+1):
            j = 1
            while i+j*j <=n:
                dp[i+j*j] = min(dp[i]+1, dp[i+j*j])
                j += 1
        return dp[n]
if __name__ == "__main__":
    sol = Solution()
    print sol.numSquares(5673)

File: downloads/code/LeetCode/Java/Perfect-Squares.java
public class Solution {
    public int numSquares(int n) {
        int[] dp = new int[n+1];
        for(int i=1; i<=n; i++)
            dp[i] = Integer.MAX_VALUE;
        // init 所有完全平方数
        for(int i=1; i<=n; i++){
            if(i*i<=n) 
                dp[i*i]=1;
        }
        
     
        for(int i=1;i<=n;i++){
            for(int j=1; i+j*j<=n; j++){
                dp[i+j*j] = Math.min(dp[i]+1,dp[i+j*j]);
            }
        }
        return dp[n];
    }
}

